packages/styled-components/src/models/StyledComponent.js

```diff
@@ -145,15 +147,25 @@ function useStyledComponentImpl<Config: {}, Instance>(
 
   const isTargetTag = isTag(elementToBeCreated);
   const computedProps = attrs !== props ? { ...props, ...attrs } : props;
-  const shouldFilterProps = isTargetTag || 'as' in computedProps || 'forwardedAs' in computedProps;
+  const shouldFilterProps = shouldForwardProp || isTargetTag || 'as' in computedProps || 'forwardedAs' in computedProps;
   const propsForElement = shouldFilterProps ? {} : { ...computedProps };
 
+  // use explicit `shouldForwardProp` function if one is set, otherwise
+  // if no explicit `shouldForwardProp` set then
+  //   if we're on a tag, use `validAttr`
+  //     i.e. Don't pass through non HTML tags through to HTML elements
+  // else (we're on a component - pass through all by default)
+  //   () => true
+  const shouldReallyForwardProp = shouldForwardProp || (
```

Comment:
I think the comment here is redundant if this was a little clearer in the first place 😅 

If we separate this out a little then we can make this very clear.

- `shouldFilterProps` could be reverted and we can remove `isTargetTag` from it.
- We can introduce a separate variables: `const propFilterFn = shouldForwardProp || (isTargetTag && validAttr);`.
- `if (shouldFilterProps)` becomes `if (shouldFilterProps || propFilterFn)`
- Then `shouldReallyForwardProp(key)` (which btw has unnecessary parens) can be replaced by: `(!propFilterFn || propFilterFn(key))`
- lastly we can now rename `shouldFilterProps` to `shouldFilterComputedProps`

Reply:
done - was a little bit more involved, but essentially followed your suggestions - also didn't rename `shouldFilterProps` cos that didn't feel entirely worth it :grin: